onadata/libs/pagination.py

```diff
@@ -57,3 +62,73 @@ def paginate_queryset(self, queryset, request, view, count=None):
 
         self.request = request
         return list(self.page)
+
+
+def generate_pagination_headers(
+    request: Request, no_of_records: int, current_page_size: Union[int, str],
+    current_page: Union[int, str]
+) -> dict:
+    url = urlparse(request.build_absolute_uri())
+    base_url = f"{url.scheme}://{url.netloc}{url.path}?"
+    if url.query:
+        query_params = parse_qs(url.query)
+        query_string = None
+        for param_key, param_values in query_params.items():
+            if param_key not in ['page_size', 'page']:
+                param_value = ','.join(param_values)
+                query = f'{param_key}={param_value}&'
+                if not query_string:
+                    query_string = query
+                else:
+                    query_string += query
+
+        base_url += f'{query_string}'
+
+    next_page_url = None
+    prev_page_url = None
+    first_page_url = None
+    last_page_url = None
+    links = []
+
+    if isinstance(current_page, str):
+        try:
+            current_page = int(current_page)
+        except ValueError:
+            return
+
+    if isinstance(current_page_size, str):
+        try:
+            current_page_size = int(current_page_size)
+        except ValueError:
+            return
+
+    if (current_page * current_page_size) < no_of_records:
+        next_page_url = (
+            f"{base_url}page={current_page + 1}&"
+            f"page_size={current_page_size}")
+
+    if current_page > 1:
+        prev_page_url = (
+            f"{base_url}page={current_page - 1}"
+            f"&page_size={current_page_size}")
+
+    last_page = math.ceil(no_of_records / current_page_size)
+    if last_page != current_page and last_page != current_page + 1:
+        last_page_url = (
+            f"{base_url}page={last_page}&page_size={current_page_size}"
+        )
+
+    if current_page != 1:
+        first_page_url = (
+            f"{base_url}page=1&page_size={current_page_size}"
+        )
+
+    for rel, link in (
+            ('prev', prev_page_url),
+            ('next', next_page_url),
+            ('last', last_page_url),
+            ('first', first_page_url)):
```

Comment:
For this, I'm mostly curious, I'm not sure which data structure this is 
```code
(('prev', prev_page_url),
('next', next_page_url),
('last', last_page_url),
('first', first_page_url))
```
Why did we use this struct over an array of tuples?

Reply:
It's a "nested" tuple in some sense... not entirely sure if that's the best way to call it though. Using this since it seems easier to loop through it's tuple objects and values. I could use a struct or dict in this case but this seems simpler IMO